### PR TITLE
Implemented Infty::pow and Infty::rpow functions

### PR DESCRIPTION
**Pull request by ShikharJ**
As always, please review and suggest changes.

### CODE REVIEW

**Comment by certik** [comment]
Thanks for implementing these. Please do not forget to also add some tests.

**Comment by ShikharJ** [comment]
ping @isuruf

File: symengine/infinity.cpp [hunk 1]
```diff
@@ -143,15 +144,113 @@ RCP<const Number> Infty::div(const Number &other) const
...
+            }
+        }
+    }
 }
-// TODO
+
 RCP<const Number> Infty::rpow(const Number &other) const
 {
-    return zero;
+    if (is_a<Infty>(other)) {
```
Comment (isuruf) [inline symengine/infinity.cpp]: `rpow` doesn't need this case, it is handled in `pow`

File: symengine/tests/basic/test_infinity.cpp [hunk 2]
```diff
@@ -225,3 +225,77 @@ TEST_CASE("Division of Infinity", "[Infinity]")
...
+    // REQUIRE(n1->__str__() == "oo");
+    n1 = a->rpow(*rational(2, 5));
+    REQUIRE(eq(*n1, *zero));
+    n1 = a->rpow(*a);
+    REQUIRE(n1->__str__() == "oo");
+    n1 = a->rpow(*c);
+    REQUIRE(n1->__str__() == "zoo");
+    n1 = b->rpow(*a);
+    REQUIRE(eq(*n1, *zero));
+    n1 = b->rpow(*c);
```
Comment (isuruf) [inline symengine/tests/basic/test_infinity.cpp]: Don't use `rpow` directly. Use `c->rpow(*b)`

**Comment by ShikharJ** [comment]
@isuruf @certik The changes have been made. Please review and suggest improvements.

File: symengine/infinity.cpp [hunk 3]
```diff
@@ -143,15 +144,91 @@ RCP<const Number> Infty::div(const Number &other) const
...
+        return other.pow(*this);
```
Comment (isuruf) [inline symengine/infinity.cpp]: You should not need this at all, since `rpow` is not directly called.

File: symengine/infinity.cpp [hunk 4]
```diff
@@ -143,15 +144,91 @@ RCP<const Number> Infty::div(const Number &other) const
...
+        throw NotImplementedError(
+            "Raising Complex powers to Infty not yet implemented");
+    } else {
+        if (other.is_negative()) {
+            throw NotImplementedError("NaN module not yet implemented");
+        } else if (other.is_zero()) {
+            throw SymEngineException("Indeterminate Expression: `0 ** +- "
+                                     "unsigned Infty` encountered");
+        } else {
+            const RealDouble &s = static_cast<const RealDouble &>(other);
```
Comment (isuruf) [inline symengine/infinity.cpp]: How do you know that `other` is a `RealDouble`?

File: symengine/tests/basic/test_infinity.cpp [hunk 5]
```diff
@@ -225,3 +225,76 @@ TEST_CASE("Division of Infinity", "[Infinity]")
...
+}
+
+TEST_CASE("Powers to Infinity", "[Infinity]")
+{
+    RCP<const Infty> a = Inf;
+    RCP<const Infty> b = NegInf;
+    RCP<const Infty> c = ComplexInf;
+
+    RCP<const Number> n1;
+    n1 = a->rpow(*integer(10));
```
Comment (isuruf) [inline symengine/tests/basic/test_infinity.cpp]: Don't call `rpow` directly

File: symengine/infinity.cpp (hunk 4 above)
Comment (ShikharJ) [inline symengine/infinity.cpp]: Actually I wanted to down-cast the value held by the object into a double and compare with 1. But using 'down_cast()' is raising an error.

File: symengine/infinity.cpp [hunk 6]
```diff
@@ -143,15 +144,88 @@ RCP<const Number> Infty::div(const Number &other) const
...
+            throw NotImplementedError("NaN module not yet implemented");
+        } else if (other.is_zero()) {
+            throw SymEngineException("Indeterminate Expression: `0 ** +- "
+                                     "unsigned Infty` encountered");
+        } else {
+            if (other.is_one()) {
+                throw SymEngineException("Indeterminate Expression: `1 ** +- "
+                                         "unsigned Infty` encountered");
+            } else if (is_positive_infinity()) {
+                if (other.compare(*one) == -1) {
```
Comment (isuruf) [inline symengine/infinity.cpp]: This comparison is not mathematical, but structural. check whether `other - 1` is positive or zero or negative.

File: symengine/infinity.cpp [hunk 7]
```diff
@@ -143,15 +144,88 @@ RCP<const Number> Infty::div(const Number &other) const
...
+            }
+        }
+    }
 }
-// TODO
+
 RCP<const Number> Infty::rpow(const Number &other) const
 {
-    return zero;
+    if (is_a<Complex>(other)) {
```
Comment (isuruf) [inline symengine/infinity.cpp]: Check for `ComplexDouble` and `ComplexMPC` here as well.

**Comment by ShikharJ** [comment]
Ping @isuruf.

**Pull request merged**